### PR TITLE
crosscluster: update DLQ tests

### PR DESCRIPTION
**Pull request by azhu-crl**
Changes in this PR:
- Added testing knob to inject failure
- Added e2e DLQ test
- Refactoring

Epic: none
Release note: none

### CODE REVIEW

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 1]
```diff
@@ -1120,6 +1120,32 @@ func WaitUntilReplicatedTime(
 	})
 }
 
+func WaitForDLQProgress(t *testing.T, db *sqlutils.SQLRunner, tableName string) {
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1123]: nit: name could be more descriptive: `WaitForNonEmptyDLQ`

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 2]
```diff
@@ -1120,6 +1120,32 @@ func WaitUntilReplicatedTime(
...
+		var numRows int
+		db.QueryRow(t, query).Scan(&numRows)
+		if numRows == 0 {
+			return errors.Newf("DLQ table is empty")
+		}
+		return nil
+	})
+}
+
+func getTableDescriptorID(t *testing.T, db *sqlutils.SQLRunner, dbName, tableName string) int32 {
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1136]: nit: use the `sqlutils.QueryTableID` helper

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 3]
```diff
@@ -1568,3 +1594,148 @@ func TestUserPrivileges(t *testing.T) {
 		jobutils.WaitForJobToPause(t, dbA, jobAID)
 	})
 }
+
+func TestDLQ(t *testing.T) {
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1598]: any reason you put this in logical_replication_job_test instead of in dead_letter_queue_test ?

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 4]
```diff
@@ -1568,3 +1594,148 @@ func TestUserPrivileges(t *testing.T) {
...
+	defer cleanup()
+	dbBURL, cleanupB := s.PGUrl(t, serverutils.DBName("b"))
+	defer cleanupB()
+
+	var (
+		jobAID jobspb.JobID
+		jobBID jobspb.JobID
+	)
+	dbA.QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE tab ON $1 INTO TABLE tab WITH DEFAULT FUNCTION = 'dlq'", dbBURL.String()).Scan(&jobAID)
+	dbB.QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE tab ON $1 INTO TABLE tab WITH DEFAULT FUNCTION = 'dlq'", dbAURL.String()).Scan(&jobBID)
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1632]: why do we need a bidirectional replication stream for this test?

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 5]
```diff
@@ -1568,3 +1594,148 @@ func TestUserPrivileges(t *testing.T) {
 		jobutils.WaitForJobToPause(t, dbA, jobAID)
 	})
 }
+
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1597]: nit: this test could use a more descriptive name. Perhaps TestEndToEndDLQ and a short sentence description?

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 6]
```diff
@@ -1568,3 +1594,148 @@ func TestUserPrivileges(t *testing.T) {
...
+			{
+				"hello, again", "1",
+			},
+			{
+				"potato", "2",
+			},
+		},
+	)
+
+	dbA.Exec(t, "CANCEL JOB $1", jobAID.String())
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1736]: why do we need to cancel the job? Faster tests are better tests 🏎️

File: pkg/sql/exec_util.go [hunk 7]
```diff
@@ -1865,6 +1865,8 @@ type StreamingTestingKnobs struct {
 	SkipSpanConfigReplication bool
 
 	SpanConfigRangefeedCacheKnobs *rangefeedcache.TestingKnobs
+
+	GetFailureRate func() uint32
```
Comment (msbutler) [inline pkg/sql/exec_util.go:1869]: this can just be an int, instead of a func that returns an int, yeah?

**Review by msbutler — COMMENTED** [review]
looks good! mostly nits

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go (hunk 3 above)
Comment (azhu-crl) [reply pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1598]: I thought that `logical_replication_job_test`  might be a better place for e2e tests but it doesn't test any ldr job functionalities. Moved the test to `dead_letter_queue_test` instead

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go (hunk 4 above)
Comment (azhu-crl) [reply pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1632]: good question, I think uni-directional stream should suffice

File: pkg/ccl/crosscluster/logical/lww_kv_processor.go [hunk 8]
```diff
@@ -95,10 +95,6 @@ var originID1Options = &kvpb.WriteOptions{OriginID: 1}
 func (p *kvRowProcessor) ProcessRow(
 	ctx context.Context, txn isql.Txn, keyValue roachpb.KeyValue, prevValue roachpb.Value,
 ) (batchStats, error) {
-	if err := p.injectFailure(); err != nil {
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/lww_kv_processor.go:98]: nit: why was this moved a couple lines down?

**Review by msbutler — APPROVED** [review]

File: pkg/ccl/crosscluster/logical/lww_kv_processor.go (hunk 8 above)
Comment (azhu-crl) [reply pkg/ccl/crosscluster/logical/lww_kv_processor.go:98]: I moved it after `p.lastRow = row` bc we want to be able to read the decoded row from the processor for the dlq test

Comment (msbutler) [reply pkg/ccl/crosscluster/logical/lww_kv_processor.go:98]: ahhhh. ty for explaining for the dull

**Comment by azhu-crl** [comment]
TFTR!

bors r=msbutler
